vortex-expr/src/forms/nnf.rs

```diff
@@ -0,0 +1,151 @@
+use vortex_error::VortexResult;
+
+use crate::traversal::{FoldChildren, FoldDown, FoldUp, Folder, Node as _};
+use crate::{not, BinaryExpr, ExprRef, Not, Operator};
+
+/// Return an equivalent expression in Negative Normal Form (NNF).
+///
+/// In NNF, [Not] expressions may only contain terminal nodes such as [Literal] or [Column]. They
+/// *may not* contain [BinaryExpr], [Not], etc.
+///
+/// # Examples
+///
+/// Double negation is removed entirely:
+///
+/// ```
+/// use vortex_expr::{not, col};
+/// use vortex_expr::forms::nnf::nnf;
+///
+/// let double_negation = not(not(col("a")));
+/// let nnfed = nnf(double_negation).unwrap();
+/// assert_eq!(&nnfed, &col("a"));
+/// ```
+///
+/// Triple negation becomes single negation:
+///
+/// ```
+/// use vortex_expr::{not, col};
+/// use vortex_expr::forms::nnf::nnf;
+///
+/// let triple_negation = not(not(not(col("a"))));
+/// let nnfed = nnf(triple_negation).unwrap();
+/// assert_eq!(&nnfed, &not(col("a")));
+/// ```
+///
+/// Negation at a high-level is pushed to the leaves, likely increasing the total number nodes:
+///
+/// ```
+/// use vortex_expr::{not, col, and, or};
+/// use vortex_expr::forms::nnf::nnf;
+///
+/// assert_eq!(
+///     &nnf(not(and(col("a"), col("b")))).unwrap(),
+///     &or(not(col("a")), not(col("b")))
+/// );
+/// ```
+///
+/// In Vortex, NNF is extended beyond simple Boolean operators to any Boolean-valued operator:
+///
+/// ```
+/// use vortex_expr::{not, col, and, or, lt, lit, gt_eq};
+/// use vortex_expr::forms::nnf::nnf;
+///
+/// assert_eq!(
+///     &nnf(not(and(gt_eq(col("a"), lit(3)), col("b")))).unwrap(),
+///     &or(lt(col("a"), lit(3)), not(col("b")))
+/// );
+/// ```
+pub fn nnf(expr: ExprRef) -> VortexResult<ExprRef> {
+    let mut visitor = NNFVisitor::default();
+    Ok(expr.transform_with_context(&mut visitor, false)?.result())
+}
+
+#[derive(Default)]
+struct NNFVisitor {}
+
+impl Folder for NNFVisitor {
+    type NodeTy = ExprRef;
+    type Out = ExprRef;
+    type Context = bool;
+
+    fn visit_down(
+        &mut self,
+        node: &ExprRef,
+        negating: bool,
+    ) -> VortexResult<FoldDown<ExprRef, bool>> {
+        let node_any = node.as_any();
+        if node_any.downcast_ref::<Not>().is_some() {
+            return Ok(FoldDown::Continue(!negating));
+        } else if let Some(binary_expr) = node_any.downcast_ref::<BinaryExpr>() {
+            match binary_expr.op() {
+                Operator::And | Operator::Or => {
+                    return Ok(FoldDown::Continue(negating));
+                }
+                _ => {}
+            }
+        }
+
+        Ok(FoldDown::SkipChildren)
```

Comment:
In theory we could have ‘(a and b) == c’ if c is boolean which is ‘(a and b) nxor c’ which I guess we could simplify?

Reply:
That's a good point.

Our expressions are untyped though, so I can't do this unless NNF is provided with a type binding environment.

Reply:
Agreed, but data-fusion uses typing to simplify further, so maybe we can in the future.